runtime/controller/watch.go

```diff
@@ -0,0 +1,76 @@
+/*
+Copyright 2023 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"github.com/spf13/pflag"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	flagWatchWatchAllNamespaces = "watch-all-namespaces"
+	flagWatchLabelSelector      = "watch-label-selector"
+)
+
+// WatchOptions defines the configurable options for reconciler resources watcher.
+type WatchOptions struct {
+	// AllNamespaces defines the watch filter at namespace level.
+	// If set to false, the reconciler will only watch the runtime namespace for resource changes.
+	AllNamespaces bool
+
+	// LabelSelector defines the watch filter based on matching label expressions.
+	// When set, the reconciler will only watch for changes of those resources with matching labels.
+	// Docs: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#list-and-watch-filtering.
+	LabelSelector string
+}
+
+// BindFlags will parse the given pflag.FlagSet for the controller and
+// set the WatchOptions accordingly.
+func (o *WatchOptions) BindFlags(fs *pflag.FlagSet) {
+	fs.BoolVar(&o.AllNamespaces, flagWatchWatchAllNamespaces, true,
+		"Watch for resources in all namespaces, if set to false it will only watch the runtime namespace.")
+	fs.StringVar(&o.LabelSelector, flagWatchLabelSelector, "",
+		"Watch for resources with matching labels e.g. 'sharding.fluxcd.io/shard=shard1'.")
+}
+
+// GetWatchLabelSelector parses the label selector option from WatchOptions and returns the label selector.
+func GetWatchLabelSelector(opts WatchOptions) (*metav1.LabelSelector, error) {
+	if opts.LabelSelector == "" {
+		return nil, nil
+	}
+
+	return metav1.ParseToLabelSelector(opts.LabelSelector)
```

Comment:
Making this return the result of `metav1.LabelSelectorAsSelector` instead would save us from doing another error catch in every controller.

Reply:
Or alternative: add another `func GetWatchSelector(opts WatchOptions) (labels.Selector, error)` which wraps this.